### ISSUE TITLE
Remote host "fall-back" strategy 

### ISSUE THREAD

**Issue by gaow** [vatlab/SOS#608]
I've being running my pipeline remotely but so far I've got way more failure than success, for not very obvious reasons. A disappointing fact is that it does not seem SoS is able to monitor job status well. Perhaps screenshot below can show what I'm experiencing:

![2017-05-15-13-30-29_scrot](https://cloud.githubusercontent.com/assets/917985/26072873/bea9f79c-3972-11e7-919d-905b7a71d586.png)

Terminal above is from my local computer and below is the remote computer. Apparently via `showq` command there is no job running under my account (in fact they all have failed). Yet `sos status` command still reports `submitted`/`completed`; therefore the screen above keeps showing the same info like a zombie. My `status_cmd` setting is `squeue --job {job_id}` and as seen from the screen shot when I run it directly there is an error message, so I'm a bit confused why the monitor does not quit on error. 

However locating the `.sh` in `~/.sos` and submit that file seems to work.

I understand it is very hard to eyeball the bug here. Using more verbosity may help but not always. I'm wondering if we can have a fall-back strategy, ie in this situation can SoS provide a way to give me all the relevant `.sh` files otherwise found in `~/.sos` and allow me to `qsub` each of them? I'm thinking some `dryrun` mode can provide this so that I can run these specific `sh` over and over again until all are executed, then finally run the original `sos` command to move on. This is a bit hacky but I think it is very difficult to ensure everything always works.

**Comment by gaow** [comment]
I should add that the only way to tell if my jobs are successful is to run, on my SLURM system, `sacct`, which shows you the job status of recently submitted jobs:

```
27327547     PFA_vanil+  mstephens pi-msteph+          4     FAILED      1:0 
27327547.ba+      batch            pi-msteph+          4     FAILED      1:0 
27327548     PFA_vanil+  mstephens pi-msteph+          4     FAILED      1:0 
27327548.ba+      batch            pi-msteph+          4     FAILED      1:0 
27327550     PFA_vanil+  mstephens pi-msteph+          4     FAILED      1:0 
27327550.ba+      batch            pi-msteph+          4     FAILED      1:0 
27328521     PFA_vanil+  mstephens pi-msteph+          4  COMPLETED      0:0 
27328521.ba+      batch            pi-msteph+          4  COMPLETED      0:0 
27328529     PFA_vanil+  mstephens pi-msteph+          4  COMPLETED      0:0 
27328529.ba+      batch            pi-msteph+          4  COMPLETED      0:0 
27328530     PFA_vanil+  mstephens pi-msteph+          4  COMPLETED      0:0 
27328530.ba+      batch            pi-msteph+          4  COMPLETED      0:0 
27328531     PFA_vanil+  mstephens pi-msteph+          4    RUNNING      0:0 
```

This has to be executed on the remote host usually the login node. I'm also confused by why there seems to be 2 jobs for a task:

```
27328530.ba+      batch            pi-msteph+          4  COMPLETED      0:0 
27328531     PFA_vanil+  mstephens pi-msteph+          4    RUNNING      0:0 
```
A SLURM artifact or SoS design? In any case, I'm just saying that there is something `sos task` cannot  detect. The question is should we try really hard to make sure it works on all systems, or provide some generic fall back solution.

**Comment by BoPeng** [comment]
I think the problem here is that

1. SoS submits the job to the queue, change the status to `submitted`. This state is marked by the existence and the execution of a `.sh` file, and the job is not yet running.
2. SoS expects the status changed to
  * `running` if the pulse thread starts from `sos execute`.
  * then `failed` if the pulse file exists but sos quits with a `.res` file with error code.
  * then `aborted` if the pulse file stops.
3. However, your `slurm` script seemed failed without executing the `sos` command, so it fails even before a pulse thread is started, so to the eyes of sos nothing has changed compared to the `submitted` status.

**Comment by gaow** [comment]
I think your diagnostic is exactly right. But even with `-v3` I was not able to see what exactly has caused the failure. Submitting it over and over again resolves it. In my job template the only command I run before `sos task` is `mkdir -p  ... `

**Comment by BoPeng** [comment]
So you meant the `ssh qsub` failed but local `qsub task.sh` succeeded? In theory as long as the job has been submitted (with a job id), it does not matter how the job is submitted.

**Comment by gaow** [comment]
Exactly. Also running `sos` again later work. It's just some cluster mystery I run into from time to time. I was just complaining that it is not easy to ensure all is correct and very likely SoS will be infinitely monitoring jobs that have long gone. In my case though lacking generality, I can imaging somehow SoS takes `sacct` info, parse it, and use it as an additional measure to check on job status (because this is how i can tell my job failed). 

Still my `status_cmd` shows no jobs running but SoS does not handle that information ... is this because my status check command is not exactly empty string?
```             
JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
```

**Comment by BoPeng** [comment]
OK, I tried to improve the detection of `submitted` status, please test.

**Comment by BoPeng** [comment]
I know you got other problems but let us get this `submitted` status fixed first.

**Comment by gaow** [comment]
Sorry for the delayed feedback. Was working with a colleague until just now. We've updated to the improved patch at around 4pm and we've been using it since. I think it is able to prevent some very early stage issue better now, which is a bit improvement! But there are still smaller issues, some of which might involve other complications. Maybe I'd like to check this one quickly here: for `Py_Module` `R_library` etc does it check both on remote and on localhost?  Does not look like the case as I ended up having to install all new packages manually on the cluster ...
